src/bundle.rs

```diff
@@ -354,6 +358,39 @@ impl<T: Authorization, V> Bundle<T, V> {
             .map(|a| a.to_instance(self.flags, self.anchor))
             .collect()
     }
+
+    /// Perform trial decryption of each action in the bundle with each of the
+    /// specified incoming viewing keys, and return the decrypted note contents
+    /// along with the index of the action from which it was derived.
+    pub fn decrypt_outputs_for_keys(
+        &self,
+        keys: &[IncomingViewingKey],
+    ) -> Vec<(usize, IncomingViewingKey, Note, Address, [u8; 512])> {
+        self.actions
+            .iter()
+            .enumerate()
+            .filter_map(|(idx, action)| {
+                let domain = OrchardDomain::for_action(action);
+                keys.iter().find_map(move |ivk| {
+                    try_note_decryption(&domain, ivk, action)
+                        .map(|(n, a, m)| (idx, ivk.clone(), n, a, m))
```

Comment:
```suggestion
                        .map(|(note, recipient, memo)| (idx, ivk.clone(), note, recipient, memo))
```

Reply:
Consider simplifying the `try_note_decryption` API: it's redundant to have both `note` and `recipient` in the returned tuple because `recipient` can be derived from `note` (even pre-ZIP 212).

Reply:
We need the `recipient` argument because the [Sapling note type in `zcash_primitives`](https://github.com/zcash/librustzcash/blob/6e39480c589bcb1b548c507aa96575368e72952a/zcash_primitives/src/sapling.rs#L406-L407) only contains `g_d`, not `d` (because the type was originally written as a circuit type, which only defines a note as containing `g_d`, and later was generalised without being refactored). I've opened https://github.com/zcash/librustzcash/issues/455 to track the removal of this API (which has previously only been documented in the `zcash_note_encryption` source code).